File: src/main/java/dev/confusedalex/thegoldeconomy/EconomyImplementer.java
package dev.confusedalex.thegoldeconomy;

import net.milkbowl.vault.economy.Economy;
import net.milkbowl.vault.economy.EconomyResponse;
import org.bukkit.Bukkit;
import org.bukkit.OfflinePlayer;
import org.bukkit.entity.Player;

import java.util.List;
import java.util.Optional;
import java.util.ResourceBundle;
import java.util.UUID;

public class EconomyImplementer implements Economy {
  TheGoldEconomy plugin;
  Bank bank;
  Converter converter;
  ResourceBundle bundle;
  Util util;

  public EconomyImplementer(TheGoldEconomy plugin, ResourceBundle bundle, Util util) {
    this.plugin = plugin;
    this.bundle = bundle;
    this.util = util;
    converter = new Converter(this, bundle);
    bank = new Bank(converter);
  }

  @Override
  public boolean isEnabled() {
    return true;
  }

  @Override
  public String getName() {
    return "TheGoldEconomy";
  }

  @Override
  public boolean hasBankSupport() {
    return false;
  }

  @Override
  public int fractionalDigits() {
    return 0;
  }

  @Override
  public String format(double v) {
    return v + " Gold";
  }

  @Override
  public String currencyNamePlural() {
    return "Gold";
  }

  @Override
  public String currencyNameSingular() {
    return "Gold";
  }

  @Override
  public boolean hasAccount(String s) {
    if (util.isOfflinePlayer(s).isPresent()) return true;
    if (bank.getFakeAccounts().containsKey(s)) return true;

    bank.setFakeAccountBalance(s, 0);
    return true;
  }

  @Override
  public boolean hasAccount(OfflinePlayer offlinePlayer) {
    return true;
  }

  @Override
  public boolean hasAccount(String s, String s1) {
    if (util.isOfflinePlayer(s).isPresent()) return true;
    if (bank.getFakeAccounts().containsKey(s)) return true;

    bank.setFakeAccountBalance(s, 0);
    return true;
  }

  @Override
  public boolean hasAccount(OfflinePlayer offlinePlayer, String s) {
    return true;
  }

  @Override
  public double getBalance(String s) {
    try {
      UUID uuid = UUID.fromString(s);
      if (Bukkit.getPlayer(uuid) != null) return bank.getTotalPlayerBalance(uuid);
    } catch (IllegalArgumentException e) {
      // String is not UUID
    }
    Optional<OfflinePlayer> playerOptional = util.isOfflinePlayer(s);
      return playerOptional.map(offlinePlayer -> bank.getTotalPlayerBalance(offlinePlayer.getUniqueId())).orElseGet(() -> bank.getFakeBalance(s));
  }

  @Override
  public double getBalance(OfflinePlayer offlinePlayer) {
    if (offlinePlayer != null) return bank.getTotalPlayerBalance(offlinePlayer.getUniqueId());
    return 0;
  }

  @Override
  public double getBalance(String s, String s1) {
    try {
      UUID uuid = UUID.fromString(s);
      if (Bukkit.getPlayer(uuid) != null) return bank.getTotalPlayerBalance(uuid);
    } catch (IllegalArgumentException e) {
      // String is not UUID
    }
    if (util.isOfflinePlayer(s).isPresent())
      return bank.getTotalPlayerBalance(Bukkit.getOfflinePlayer(s).getUniqueId());
    return bank.getFakeBalance(s);
  }

  @Override
  public double getBalance(OfflinePlayer offlinePlayer, String s) {
    if (offlinePlayer != null) return bank.getTotalPlayerBalance(offlinePlayer.getUniqueId());
    return 0;
  }

  @Override
  public boolean has(String s, double v) {
    if (util.isOfflinePlayer(s).isPresent())
      return v < bank.getTotalPlayerBalance(Bukkit.getOfflinePlayer(s).getUniqueId());
    else return v < bank.getFakeBalance(s);
  }

  @Override
  public boolean has(OfflinePlayer offlinePlayer, double v) {
    return v < bank.getTotalPlayerBalance(offlinePlayer.getUniqueId());
  }

  @Override
  public boolean has(String s, String s1, double v) {
    if (util.isOfflinePlayer(s).isPresent())
      return v < bank.getTotalPlayerBalance(Bukkit.getOfflinePlayer(s).getUniqueId());
    else return v < bank.getFakeBalance(s);
  }

  @Override
  public boolean has(OfflinePlayer offlinePlayer, String s, double v) {
    return v < bank.getTotalPlayerBalance(offlinePlayer.getUniqueId());
  }

  @Override
  public EconomyResponse withdrawPlayer(String s, double amount) {
    int oldBalance = 0;

    // if amount is negative return
    if (amount < 0) return new EconomyResponse(amount, oldBalance, EconomyResponse.ResponseType.FAILURE, "error");

    Optional<OfflinePlayer> playerOptional = util.isOfflinePlayer(s);
    if (playerOptional.isPresent()) {
      OfflinePlayer offlinePlayer = playerOptional.get();
      UUID uuid = offlinePlayer.getUniqueId();

      // if player is online
      if (offlinePlayer.isOnline()) {
        Player player = offlinePlayer.getPlayer();

        if (player == null)
          return new EconomyResponse(amount, oldBalance, EconomyResponse.ResponseType.FAILURE, "error");

        // get balance and InventoryValue from Player
        int oldBankBalance = bank.getAccountBalance(uuid);
        int oldInventoryBalance = converter.getInventoryValue(player);


        // If balance + InventoryValue is < amount, return
        if (amount > oldBankBalance + oldInventoryBalance)
          return new EconomyResponse(amount, oldBalance, EconomyResponse.ResponseType.FAILURE, "Not enough Money!");
        // If bank balances is enough to cover amount
        if (oldBankBalance - amount > 0) {
          bank.setAccountBalance(uuid, (int) (oldBankBalance - amount));
          return new EconomyResponse(amount, (oldBankBalance - amount), EconomyResponse.ResponseType.SUCCESS, "");
        } else {
          // Set balance to 0 and cover rest of the costs with Inventory Funds
          int diff = (int) (amount - oldBankBalance);
          bank.setAccountBalance(uuid, 0);
          converter.remove(player, diff);

          return new EconomyResponse(amount, oldInventoryBalance - amount, EconomyResponse.ResponseType.SUCCESS, "");
        }
      } else {
        // When player is offline
        oldBalance = bank.getTotalPlayerBalance(uuid);
        int newBalance = (int) (oldBalance - amount);
        bank.setAccountBalance(uuid, newBalance);
        return new EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "");
      }
    } else {
      oldBalance = bank.getFakeBalance(s);
      int newBalance = (int) (oldBalance - amount);
      bank.setFakeAccountBalance(s, newBalance);
      return new EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "");
    }
  }

  @Override
  public EconomyResponse withdrawPlayer(OfflinePlayer offlinePlayer, double amount) {
    UUID uuid = offlinePlayer.getUniqueId();
    Player player;
    int oldBalance = 0;

    // if amount is negative return
    if (amount < 0) return new EconomyResponse(amount, oldBalance, EconomyResponse.ResponseType.FAILURE, "error");

    // if player is online
    if (offlinePlayer.isOnline()) {
      player = offlinePlayer.getPlayer();

      if (player == null)
        return new EconomyResponse(amount, oldBalance, EconomyResponse.ResponseType.FAILURE, "error");

      // get Balance and InventoryValue
      int oldBankBalance = bank.getAccountBalance(uuid);
      int oldInventoryBalance = converter.getInventoryValue(player);

      // If balance + InventoryValue is < amount, return
      if (amount > oldBankBalance + oldInventoryBalance)
        return new EconomyResponse(amount, oldBalance, EconomyResponse.ResponseType.FAILURE, "error");
      // If bank balances is enough to cover amount
      if (oldBankBalance - amount > 0) {
        bank.setAccountBalance(uuid, (int) (oldBankBalance - amount));
        return new EconomyResponse(amount, (oldBankBalance - amount), EconomyResponse.ResponseType.SUCCESS, "");
      } else {
        // Set balance to 0 and cover rest of the costs with Inventory Funds
        int diff = (int) (amount - oldBankBalance);
        bank.setAccountBalance(uuid, 0);
        converter.remove(player, diff);
        return new EconomyResponse(amount, oldInventoryBalance - amount, EconomyResponse.ResponseType.SUCCESS, "");
      }
    } else {
      // if offline or fakeAccount
      oldBalance = bank.getTotalPlayerBalance(uuid);
      int newBalance = (int) (oldBalance - amount);
      bank.setAccountBalance(uuid, newBalance);

      return new EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "");
    }
  }

  @Override
  public EconomyResponse withdrawPlayer(String s, String s1, double amount) {
    int oldBalance = 0;

    // if amount is negative return
    if (amount < 0) return new EconomyResponse(amount, oldBalance, EconomyResponse.ResponseType.FAILURE, "error");

    Optional<OfflinePlayer> playerOptional = util.isOfflinePlayer(s);
    if (playerOptional.isPresent()) {
      OfflinePlayer offlinePlayer = playerOptional.get();
      UUID uuid = offlinePlayer.getUniqueId();

      // if player is online
      if (offlinePlayer.isOnline()) {
        Player player = offlinePlayer.getPlayer();

        if (player == null)
          return new EconomyResponse(amount, oldBalance, EconomyResponse.ResponseType.FAILURE, "error");

        // get balance and InventoryValue from Player
        int oldBankBalance = bank.getAccountBalance(uuid);
        int oldInventoryBalance = converter.getInventoryValue(player);


        // If balance + InventoryValue is < amount, return
        if (amount > oldBankBalance + oldInventoryBalance)
          return new EconomyResponse(amount, oldBalance, EconomyResponse.ResponseType.FAILURE, "Not enough Money!");
        // If bank balances is enough to cover amount
        if (oldBankBalance - amount > 0) {
          bank.setAccountBalance(uuid, (int) (oldBankBalance - amount));
          return new EconomyResponse(amount, (oldBankBalance - amount), EconomyResponse.ResponseType.SUCCESS, "");
        } else {
          // Set balance to 0 and cover rest of the costs with Inventory Funds
          int diff = (int) (amount - oldBankBalance);
          bank.setAccountBalance(uuid, 0);
          converter.remove(player, diff);

          return new EconomyResponse(amount, oldInventoryBalance - amount, EconomyResponse.ResponseType.SUCCESS, "");
        }
      } else {
        // When player is offline
        oldBalance = bank.getTotalPlayerBalance(uuid);
        int newBalance = (int) (oldBalance - amount);
        bank.setAccountBalance(uuid, newBalance);
        return new EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "");
      }
    } else {
      oldBalance = bank.getFakeBalance(s);
      int newBalance = (int) (oldBalance - amount);
      bank.setFakeAccountBalance(s, newBalance);
      return new EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "");
    }
  }

  @Override
  public EconomyResponse withdrawPlayer(OfflinePlayer offlinePlayer, String s, double amount) {
    UUID uuid = offlinePlayer.getUniqueId();
    Player player;
    int oldBalance = 0;

    // if amount is negative return
    if (amount < 0) return new EconomyResponse(amount, oldBalance, EconomyResponse.ResponseType.FAILURE, "error");

    // if player is online
    if (offlinePlayer.isOnline()) {
      player = offlinePlayer.getPlayer();

      if (player == null)
        return new EconomyResponse(amount, oldBalance, EconomyResponse.ResponseType.FAILURE, "error");

      // get Balance and InventoryValue
      int oldBankBalance = bank.getAccountBalance(uuid);
      int oldInventoryBalance = converter.getInventoryValue(player);

      // If balance + InventoryValue is < amount, return
      if (amount > oldBankBalance + oldInventoryBalance)
        return new EconomyResponse(amount, oldBalance, EconomyResponse.ResponseType.FAILURE, "error");
      // If bank balances is enough to cover amount
      if (oldBankBalance - amount > 0) {
        bank.setAccountBalance(uuid, (int) (oldBankBalance - amount));
        return new EconomyResponse(amount, (oldBankBalance - amount), EconomyResponse.ResponseType.SUCCESS, "");
      } else {
        // Set balance to 0 and cover rest of the costs with Inventory Funds
        int diff = (int) (amount - oldBankBalance);
        bank.setAccountBalance(uuid, 0);
        converter.remove(player, diff);

        return new EconomyResponse(amount, oldInventoryBalance - amount, EconomyResponse.ResponseType.SUCCESS, "");
      }
    } else {
      // if offline
      oldBalance = bank.getTotalPlayerBalance(uuid);
      int newBalance = (int) (oldBalance - amount);
      bank.setAccountBalance(uuid, newBalance);

      return new EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "");
    }
  }

  @Override
  public EconomyResponse depositPlayer(String s, double amount) {
    int oldBalance = 0;

    // If amount is negative -> return
    if (amount < 0) return new EconomyResponse(amount, oldBalance, EconomyResponse.ResponseType.FAILURE, "error");

    Optional<OfflinePlayer> playerOptional = util.isOfflinePlayer(s);
    if (playerOptional.isPresent()) {
      OfflinePlayer player = playerOptional.get();
      UUID uuid = player.getUniqueId();

      // Getting balance and calculating new Balance
      oldBalance = bank.getAccountBalance(uuid);
      int newBalance = (int) (oldBalance + amount);
      bank.setAccountBalance(uuid, newBalance);
      return new EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "");
    } else {
      oldBalance = bank.getFakeBalance(s);
      int newBalance = (int) (oldBalance + amount);
      bank.setFakeAccountBalance(s, newBalance);
      return new EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "");
    }
  }

  @Override
  public EconomyResponse depositPlayer(OfflinePlayer offlinePlayer, double amount) {
    UUID uuid = offlinePlayer.getUniqueId();
    int oldBalance = bank.getAccountBalance(uuid);
    int newBalance = (int) (amount + oldBalance);

    // If amount is negative -> return
    if (amount < 0) return new EconomyResponse(amount, oldBalance, EconomyResponse.ResponseType.FAILURE, "error");

    bank.setAccountBalance(uuid, newBalance);
    return new EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "");
  }

  @Override
  public EconomyResponse depositPlayer(String s, String s1, double amount) {
    int oldBalance = 0;

    // If amount is negative -> return
    if (amount < 0) return new EconomyResponse(amount, oldBalance, EconomyResponse.ResponseType.FAILURE, "error");

    Optional<OfflinePlayer> playerOptional = util.isOfflinePlayer(s);
    if (playerOptional.isPresent()) {
      OfflinePlayer player = playerOptional.get();
      UUID uuid = player.getUniqueId();

      // Getting balance and calculating new Balance
      oldBalance = bank.getAccountBalance(uuid);
      int newBalance = (int) (oldBalance + amount);
      bank.setAccountBalance(uuid, newBalance);
      return new EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "");
    } else {
      oldBalance = bank.getFakeBalance(s);
      int newBalance = (int) (oldBalance + amount);
      bank.setFakeAccountBalance(s, newBalance);
      return new EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "");
    }
  }

  @Override
  public EconomyResponse depositPlayer(OfflinePlayer offlinePlayer, String s, double amount) {
    UUID uuid = offlinePlayer.getUniqueId();
    int oldBalance = bank.getAccountBalance(uuid);
    int newBalance = (int) (amount + oldBalance);

    // If amount is negative -> return
    if (amount < 0) return new EconomyResponse(amount, oldBalance, EconomyResponse.ResponseType.FAILURE, "error");

    bank.setAccountBalance(uuid, newBalance);
    return new EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "");
  }

  @Override
  public EconomyResponse createBank(String s, String s1) {
    return null;
  }

  @Override
  public EconomyResponse createBank(String s, OfflinePlayer offlinePlayer) {
    return null;
  }

  @Override
  public EconomyResponse deleteBank(String s) {
    return null;
  }

  @Override
  public EconomyResponse bankBalance(String s) {
    return null;
  }

  @Override
  public EconomyResponse bankHas(String s, double v) {
    return null;
  }

  @Override
  public EconomyResponse bankWithdraw(String s, double v) {
    return null;
  }

  @Override
  public EconomyResponse bankDeposit(String s, double v) {
    return null;
  }

  @Override
  public EconomyResponse isBankOwner(String s, String s1) {
    return null;
  }

  @Override
  public EconomyResponse isBankOwner(String s, OfflinePlayer offlinePlayer) {
    return null;
  }

  @Override
  public EconomyResponse isBankMember(String s, String s1) {
    return null;
  }

  @Override
  public EconomyResponse isBankMember(String s, OfflinePlayer offlinePlayer) {
    return null;
  }

  @Override
  public List<String> getBanks() {
    return null;
  }

  @Override
  public boolean createPlayerAccount(String s) {
    return false;
  }

  @Override
  public boolean createPlayerAccount(OfflinePlayer offlinePlayer) {
    return false;
  }

  @Override
  public boolean createPlayerAccount(String s, String s1) {
    return false;
  }

  @Override
  public boolean createPlayerAccount(OfflinePlayer offlinePlayer, String s) {
    return false;
  }
}

File: src/main/kotlin/EconomyImplementer.kt
package thegoldeconomy

import net.milkbowl.vault.economy.Economy
import net.milkbowl.vault.economy.EconomyResponse
import org.bukkit.Bukkit
import org.bukkit.OfflinePlayer
import org.bukkit.entity.Player
import java.util.*

class EconomyImplementer(private val util: Util, private val bank: Bank, private val converter: Converter): Economy {

    override fun isEnabled(): Boolean {
        return true
    }

    override fun getName(): String {
        return "TheGoldEconomy"
    }

    override fun hasBankSupport(): Boolean {
        return false
    }

    override fun fractionalDigits(): Int {
        return 0
    }

    override fun format(v: Double): String {
        return "$v Gold"
    }

    override fun currencyNamePlural(): String {
        return "Gold"
    }

    override fun currencyNameSingular(): String {
        return "Gold"
    }

    override fun hasAccount(s: String): Boolean {
        if (util.isOfflinePlayer(s)) return true

        if (bank.fakeAccounts.containsKey(s)) return true
        else if (util.getFakeAccountsFile().contains(s)) return true

        util.getFakeAccountsFile().set(s, 0)
        bank.fakeAccounts[s] = 0
        return true
    }

    override fun hasAccount(offlinePlayer: OfflinePlayer?): Boolean {
        return true
    }

    override fun hasAccount(s: String, s1: String?): Boolean {
        if (util.isOfflinePlayer(s)) return true

        if (bank.fakeAccounts.containsKey(s)) return true
        else if (util.getFakeAccountsFile().contains(s)) return true

        util.getFakeAccountsFile().set(s, 0)
        bank.fakeAccounts[s] = 0
        return true
    }

    override fun hasAccount(offlinePlayer: OfflinePlayer, s: String): Boolean {
        return true
    }

    override fun getBalance(playerName: String): Double {
        val player = Bukkit.getPlayer(playerName)
        if (player != null) {
            return bank.getTotalPlayerBalance(player.uniqueId).toDouble()
        }

        if (util.isOfflinePlayer(playerName)) {
            return bank.getTotalPlayerBalance(Bukkit.getOfflinePlayer(playerName).uniqueId).toDouble()
        }

        return bank.getFakeBalance(playerName).toDouble()
    }

    override fun getBalance(offlinePlayer: OfflinePlayer): Double {
        return bank.getTotalPlayerBalance(offlinePlayer.uniqueId).toDouble()
    }

    override fun getBalance(playerName: String, world: String?): Double {
        val player = Bukkit.getPlayer(playerName)
        if (player != null) {
            return bank.getTotalPlayerBalance(player.uniqueId).toDouble()
        }

        if (util.isOfflinePlayer(playerName)) {
            return bank.getTotalPlayerBalance(Bukkit.getOfflinePlayer(playerName).uniqueId).toDouble()
        }

        val fb = bank.getFakeBalance(playerName)
        return fb.toDouble()
    }

    override fun getBalance(offlinePlayer: OfflinePlayer, s: String?): Double {
        return bank.getTotalPlayerBalance(offlinePlayer.uniqueId).toDouble()
    }

    override fun has(playerName: String, v: Double): Boolean {
        return if (util.isOfflinePlayer(playerName)) {
            v < bank.getTotalPlayerBalance(Bukkit.getOfflinePlayer(playerName).uniqueId)
        } else v < bank.getFakeBalance(playerName)
    }

    override fun has(offlinePlayer: OfflinePlayer, v: Double): Boolean {
        return v < bank.getTotalPlayerBalance(offlinePlayer.uniqueId)
    }

    override fun has(playerName: String, s1: String?, v: Double): Boolean {
        return if (util.isOfflinePlayer(playerName)) {
            v < bank.getTotalPlayerBalance(Bukkit.getOfflinePlayer(playerName).uniqueId)
        } else v < bank.getFakeBalance(playerName)
    }

    override fun has(offlinePlayer: OfflinePlayer, s: String?, v: Double): Boolean {
        return v < bank.getTotalPlayerBalance(offlinePlayer.uniqueId)
    }

    override fun withdrawPlayer(s: String, amount: Double): EconomyResponse {
        // If amount to withdraw is < 0
        if (amount < 0) return EconomyResponse(amount, 0.0, EconomyResponse.ResponseType.FAILURE, "Can't withdraw negative amount!")

        if (util.isOfflinePlayer(s)) {
            val offlinePlayer = Bukkit.getOfflinePlayer(s)
            // If players balance is insufficient
            if (!has(offlinePlayer, amount)) return EconomyResponse(amount, 0.0, EconomyResponse.ResponseType.FAILURE, "Not enough money")


            if (offlinePlayer.isOnline) {
                val player: Player? = offlinePlayer.player
                val uuid: UUID? = player?.uniqueId

                val oldBankBalance = bank.getAccountBalance(offlinePlayer.uniqueId)
                val oldInventoryBalance = player?.let { converter.getInventoryValue(it) }

                // If bank balances is sufficient
                if (oldBankBalance > amount) {
                    val newBalance = oldBankBalance - amount
                    uuid?.let { bank.setAccountBalance(it, (newBalance).toInt()) }
                    return EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "")
                } else {
                    val diff = amount - oldBankBalance

                    if (player != null && uuid != null) {
                        bank.setAccountBalance(uuid, 0)
                        converter.remove(player, diff.toInt())
                        return EconomyResponse(amount,
                            oldInventoryBalance?.minus(diff)!!, EconomyResponse.ResponseType.SUCCESS, "")
                    }
                }
            } else {
                val uuid = offlinePlayer.uniqueId
                val oldBalance = bank.getAccountBalance(uuid)
                val newBalance = oldBalance - amount
                bank.setAccountBalance(uuid, newBalance.toInt())
                return EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "")
            }
        } else {
            val oldBalance = bank.getFakeBalance(s)
            val newBalance = bank.getFakeBalance(s) - amount
            if (!has(s, amount)) return EconomyResponse(amount,
                oldBalance.toDouble(), EconomyResponse.ResponseType.FAILURE, "Not enough money")

            bank.setFakeAccountBalance(s, newBalance.toInt())
            return EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "")
        }
        return EconomyResponse(amount, 0.0, EconomyResponse.ResponseType.FAILURE, "")
    }

    override fun withdrawPlayer(offlinePlayer: OfflinePlayer, amount: Double): EconomyResponse? {
        // If amount to withdraw is < 0
        if (amount < 0) return EconomyResponse(amount, 0.0, EconomyResponse.ResponseType.FAILURE, "Can't withdraw negative amount!")

        // If players balance is insufficient
        if (!has(offlinePlayer, amount)) return EconomyResponse(amount, 0.0, EconomyResponse.ResponseType.FAILURE, "Not enough money")


        if (offlinePlayer.isOnline) {
            val player: Player? = offlinePlayer.player
            val uuid: UUID? = player?.uniqueId

            val oldBankBalance = bank.getAccountBalance(offlinePlayer.uniqueId)
            val oldInventoryBalance = player?.let { converter.getInventoryValue(it) }

            // If bank balances is sufficient
            if (oldBankBalance > amount) {
                val newBalance = oldBankBalance - amount
                uuid?.let { bank.setAccountBalance(it, (newBalance).toInt()) }
                return EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "")
            } else {
                val diff = amount - oldBankBalance

                if (player != null && uuid != null) {
                    bank.setAccountBalance(uuid, 0)
                    converter.remove(player, diff.toInt())
                    return EconomyResponse(amount,
                        oldInventoryBalance?.minus(diff)!!, EconomyResponse.ResponseType.SUCCESS, "")
                }
            }
        } else {
            val uuid = offlinePlayer.uniqueId
            val oldBalance = bank.getAccountBalance(uuid)
            val newBalance = oldBalance - amount
            bank.setAccountBalance(uuid, newBalance.toInt())
            return EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "")
        }
    return EconomyResponse(amount, 0.0, EconomyResponse.ResponseType.FAILURE, "")
    }

    override fun withdrawPlayer(s: String, s1: String?, amount: Double): EconomyResponse? {
        // If amount to withdraw is < 0
        if (amount < 0) return EconomyResponse(amount, 0.0, EconomyResponse.ResponseType.FAILURE, "Can't withdraw negative amount!")

        if (util.isOfflinePlayer(s)) {
            val offlinePlayer = Bukkit.getOfflinePlayer(s)
            // If players balance is insufficient
            if (!has(offlinePlayer, amount)) return EconomyResponse(amount, 0.0, EconomyResponse.ResponseType.FAILURE, "Not enough money")


            if (offlinePlayer.isOnline) {
                val player: Player? = offlinePlayer.player
                val uuid: UUID? = player?.uniqueId

                val oldBankBalance = bank.getAccountBalance(offlinePlayer.uniqueId)
                val oldInventoryBalance = player?.let { converter.getInventoryValue(it) }

                // If bank balances is sufficient
                if (oldBankBalance > amount) {
                    val newBalance = oldBankBalance - amount
                    uuid?.let { bank.setAccountBalance(it, (newBalance).toInt()) }
                    return EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "")
                } else {
                    val diff = amount - oldBankBalance

                    if (player != null && uuid != null) {
                        bank.setAccountBalance(uuid, 0)
                        converter.remove(player, diff.toInt())
                        return EconomyResponse(amount,
                            oldInventoryBalance?.minus(diff)!!, EconomyResponse.ResponseType.SUCCESS, "")
                    }
                }
            } else {
                val uuid = offlinePlayer.uniqueId
                val oldBalance = bank.getAccountBalance(uuid)
                val newBalance = oldBalance - amount
                bank.setAccountBalance(uuid, newBalance.toInt())
                return EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "")
            }
        } else {
            val oldBalance = bank.getFakeBalance(s)
            val newBalance = bank.getFakeBalance(s) - amount
            if (!has(s, amount)) return EconomyResponse(amount,
                oldBalance.toDouble(), EconomyResponse.ResponseType.FAILURE, "Not enough money")

            bank.setFakeAccountBalance(s, newBalance.toInt())
            return EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "")
        }
        return EconomyResponse(amount, 0.0, EconomyResponse.ResponseType.FAILURE, "")
    }

    override fun withdrawPlayer(offlinePlayer: OfflinePlayer, s: String?, amount: Double): EconomyResponse? {
        // If amount to withdraw is < 0
        if (amount < 0) return EconomyResponse(amount, 0.0, EconomyResponse.ResponseType.FAILURE, "Can't withdraw negative amount!")

        // If players balance is insufficient
        if (!has(offlinePlayer, amount)) return EconomyResponse(amount, 0.0, EconomyResponse.ResponseType.FAILURE, "Not enough money")


        if (offlinePlayer.isOnline) {
            val player: Player? = offlinePlayer.player
            val uuid: UUID? = player?.uniqueId

            val oldBankBalance = bank.getAccountBalance(offlinePlayer.uniqueId)
            val oldInventoryBalance = player?.let { converter.getInventoryValue(it) }

            // If bank balances is sufficient
            if (oldBankBalance > amount) {
                val newBalance = oldBankBalance - amount
                uuid?.let { bank.setAccountBalance(it, (newBalance).toInt()) }
                return EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "")
            } else {
                val diff = amount - oldBankBalance

                if (player != null && uuid != null) {
                    bank.setAccountBalance(uuid, 0)
                    converter.remove(player, diff.toInt())
                    return EconomyResponse(amount,
                        oldInventoryBalance?.minus(diff)!!, EconomyResponse.ResponseType.SUCCESS, "")
                }
            }
        } else {
            val uuid = offlinePlayer.uniqueId
            val oldBalance = bank.getAccountBalance(uuid)
            val newBalance = oldBalance - amount
            bank.setAccountBalance(uuid, newBalance.toInt())
            return EconomyResponse(amount, newBalance, EconomyResponse.ResponseType.SUCCESS, "")
        }
        return EconomyResponse(amount, 0.0, EconomyResponse.ResponseType.FAILURE, "")
    }

    override fun depositPlayer(s: String, amount: Double): EconomyResponse? {
        var oldBalance = 0

        // If amount is negative -> return
        if (amount < 0) return EconomyResponse(
            amount,
            oldBalance.toDouble(),
            EconomyResponse.ResponseType.FAILURE,
            "error"
        )
        return if (util.isOfflinePlayer(s)) {
            val player = Bukkit.getOfflinePlayer(s)
            val uuid = player.uniqueId

            // Getting balance and calculating new Balance
            oldBalance = bank.getAccountBalance(uuid)
            val newBalance = (oldBalance + amount).toInt()
            bank.setAccountBalance(uuid, newBalance)
            EconomyResponse(amount, newBalance.toDouble(), EconomyResponse.ResponseType.SUCCESS, "")
        } else {
            oldBalance = bank.getFakeBalance(s)
            val newBalance = (oldBalance + amount).toInt()
            bank.setFakeAccountBalance(s, newBalance)
            EconomyResponse(amount, newBalance.toDouble(), EconomyResponse.ResponseType.SUCCESS, "")
        }
    }

    override fun depositPlayer(offlinePlayer: OfflinePlayer, amount: Double): EconomyResponse? {
        val uuid = offlinePlayer.uniqueId
        val oldBalance: Int = bank.getAccountBalance(uuid)
        val newBalance = (amount + oldBalance).toInt()

        // If amount is negative -> return
        if (amount < 0) return EconomyResponse(
            amount,
            oldBalance.toDouble(),
            EconomyResponse.ResponseType.FAILURE,
            "error"
        )
        bank.setAccountBalance(uuid, newBalance)
        return EconomyResponse(amount, newBalance.toDouble(), EconomyResponse.ResponseType.SUCCESS, "")
    }

    override fun depositPlayer(s: String, s1: String?, amount: Double): EconomyResponse? {
        var oldBalance = 0

        // If amount is negative -> return
        if (amount < 0) return EconomyResponse(
            amount,
            oldBalance.toDouble(),
            EconomyResponse.ResponseType.FAILURE,
            "error"
        )
        return if (util.isOfflinePlayer(s)) {
            val player = Bukkit.getOfflinePlayer(s)
            val uuid = player.uniqueId

            // Getting balance and calculating new Balance
            oldBalance = bank.getAccountBalance(uuid)
            val newBalance = (oldBalance + amount).toInt()
            bank.setAccountBalance(uuid, newBalance)
            EconomyResponse(amount, newBalance.toDouble(), EconomyResponse.ResponseType.SUCCESS, "")
        } else {
            oldBalance = bank.getFakeBalance(s)
            val newBalance = (oldBalance + amount).toInt()
            bank.setFakeAccountBalance(s, newBalance)
            EconomyResponse(amount, newBalance.toDouble(), EconomyResponse.ResponseType.SUCCESS, "")
        }
    }

    override fun depositPlayer(offlinePlayer: OfflinePlayer, s: String?, amount: Double): EconomyResponse? {
        val uuid = offlinePlayer.uniqueId
        val oldBalance: Int = bank.getAccountBalance(uuid)
        val newBalance = (amount + oldBalance).toInt()

        // If amount is negative -> return
        if (amount < 0) return EconomyResponse(
            amount,
            oldBalance.toDouble(),
            EconomyResponse.ResponseType.FAILURE,
            "error"
        )
        bank.setAccountBalance(uuid, newBalance)
        return EconomyResponse(amount, newBalance.toDouble(), EconomyResponse.ResponseType.SUCCESS, "")
    }

    override fun createBank(s: String?, s1: String?): EconomyResponse? {
        return null
    }

    override fun createBank(s: String?, offlinePlayer: OfflinePlayer?): EconomyResponse? {
        return null
    }

    override fun deleteBank(s: String?): EconomyResponse? {
        return null
    }

    override fun bankBalance(s: String?): EconomyResponse? {
        return null
    }

    override fun bankHas(s: String?, v: Double): EconomyResponse? {
        return null
    }

    override fun bankWithdraw(s: String?, v: Double): EconomyResponse? {
        return null
    }

    override fun bankDeposit(s: String?, v: Double): EconomyResponse? {
        return null
    }

    override fun isBankOwner(s: String?, s1: String?): EconomyResponse? {
        return null
    }

    override fun isBankOwner(s: String?, offlinePlayer: OfflinePlayer?): EconomyResponse? {
        return null
    }

    override fun isBankMember(s: String?, s1: String?): EconomyResponse? {
        return null
    }

    override fun isBankMember(s: String?, offlinePlayer: OfflinePlayer?): EconomyResponse? {
        return null
    }

    override fun getBanks(): List<String?>? {
        return null
    }

    override fun createPlayerAccount(s: String): Boolean {
        if (util.isOfflinePlayer(s)) {
            util.createPlayerFile(Bukkit.getOfflinePlayer(s).uniqueId)
        } else {
            if (!util.getFakeAccountsFile().contains("s")) util.getFakeAccountsFile().set(s, 0)
        }
        return true
    }

    override fun createPlayerAccount(offlinePlayer: OfflinePlayer): Boolean {
        util.createPlayerFile(offlinePlayer.uniqueId)
        return true
    }

    override fun createPlayerAccount(s: String, s1: String?): Boolean {
        if (util.isOfflinePlayer(s)) {
            util.createPlayerFile(Bukkit.getOfflinePlayer(s).uniqueId)
        } else {
            if (!util.getFakeAccountsFile().contains("s")) util.getFakeAccountsFile().set(s, 0)
        }
        return true
    }

    override fun createPlayerAccount(offlinePlayer: OfflinePlayer, s: String?): Boolean {
        util.createPlayerFile(offlinePlayer.uniqueId)
        return true
    }
}